js/modules/loki_message_api.js

```diff
@@ -155,6 +177,90 @@ class LokiMessageAPI {
     log.info(`Successful storage message to ${pubKey}`);
   }
 
+  async *retrieveNextMessage(nodeUrl) {
+    const params = {
+      pubKey: ourKey,
+      lastHash: nodeData.lastHash || '',
+    };
+    const options = {
+      timeout: 40000,
+      headers: {
+        [LOKI_LONGPOLL_HEADER]: true,
+      },
+    };
+    while (true) {
+      const result = await rpc(
+        `http://${nodeUrl}`,
+        this.snodeServerPort,
+        'retrieve',
+        params,
+        options
+      );
+      if (Array.isArray(result.messages) && result.messages.length) {
+        const filteredMessages = await this.jobQueue.add(() =>
+          filterIncomingMessages(result.messages)
+        );
+        if (filteredMessages.length) {
+          yield filteredMessages;
+        }
+      }
+    }
+  }
+
+  async openConnection(callback) {
+    while (this.ourSwarmNodes.length > 0) {
+      const url = this.ourSwarmNodes.pop();
+      const successive_failures = 0;
+      while (true) {
```

Comment:
Could potentially move `successive_failures < 3` into the while guard here and move `await sleep_for(successive_failures * 1000);` to the top or something along those lines?

Reply:
Done